libsweep/src/arch/win/serial_win.cc

```diff
@@ -0,0 +1,280 @@
+#include "serial.h"
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+
+#include <windows.h>
+#include <tchar.h>
+
+namespace sweep {
+namespace serial {
+
+typedef struct error {
+  const char* what; // always literal, do not free
+} error;
+
+typedef struct device {
+  HANDLE fd;
+  OVERLAPPED ro_;
+  OVERLAPPED wo_;
+} device;
+
+// Constructor hidden from users
+static error_s error_construct(const char* what) {
+  SWEEP_ASSERT(what);
+
+  auto out = new error{what};
+  return out;
+}
+
+const char* error_message(error_s error) {
+  SWEEP_ASSERT(error);
+
+  return error->what;
+}
+
+void error_destruct(error_s error) {
+  SWEEP_ASSERT(error);
+
+  delete error;
+}
+
+static int32_t detail_get_port_number(const char* port, error_s* error) {
+  SWEEP_ASSERT(port);
+  SWEEP_ASSERT(error);
+
+  for (int i = 0; i < 99; i++) {
+    char str[10];
+    sprintf(str, "COM%d", i);
+    if (strcmp(port, str) == 0) {
+      return i;
+    }
+  }
+
+  *error = error_construct("invalid port name");
+  return -1;
+}
+
+device_s device_construct(const char* port, int32_t bitrate, error_s* error) {
+  SWEEP_ASSERT(port);
+  SWEEP_ASSERT(bitrate > 0);
+  SWEEP_ASSERT(error);
+
+  if (bitrate != 115200) {
+    *error = error_construct("baud rate is not supported");
+    return nullptr;
+  }
+
+  // read port number from the port name
+  error_s portNumError = nullptr;
+  int portNum = detail_get_port_number(port, &portNumError);
+
+  if (portNumError || portNum < 0) {
```

Comment:
Why checking < 0? Checking for error should be good enough, shouldn't it? Otherwise the interface is a bit weird.